learning-another-language/objects.md

```diff
@@ -0,0 +1,412 @@
+```js
+const testObj = {
+  someNestedObj: {
+    someOtherNum: 4,
+  }
+};
+```
+<!-- prettier-ignore-end -->
+
+We can infinitely nest objects!
+
+In order to access nested objects, we continue to _chain_ our syntax.
+
+To get `4` from accessing `testObj`, we can use [any of the following syntax](https://replit.com/@adacore/Objects-Demo#index.js):
+
+<!-- prettier-ignore-start -->
+```js
+testObj.someNestedObj.someOtherNum;
+
+testObj['someNestedObj']['someOtherNum'];
+
+testObj.someNestedObj['someOtherNum'];
+
+testObj['someNestedObj'].someOtherNum;
+```
+<!-- prettier-ignore-end -->
+
+We evaluate these lines from left to right.
+1. First, `testObj.someNestedObj` or `testObj['someNestedObj']` accesses the inner object
+1. Then, we use the syntax with `.someOtherNum` or `['someOtherNum']` on the inner object to access its own key-value pairs.
+
+Note that we _can_ mix syntaxes in one line. This might be useful in _very_ specific cases, though the rest of the time we should strive for consistency.
+
+## Functions Are Objects
+
+Functions are objects, too!
+
+Just like the object literals we've been working with, functions _also_ have key-value pairs, even though we don't normally see them.
+
+Here are some example properties that all functions have:
+
+- `name`, which is the name of the function
+- `length`, which is the number of parameters the function expects
+
+We won't normally use these properties, but they illustrate a powerful point: we can access these properties using dot notation or square bracket notation, just like any other object.
+
+[Observe this code](https://replit.com/@adacore/Objects-Demo#index.js):
+
+<!-- prettier-ignore-start -->
+```js
+const sayHelloWorld = function() {
+    console.log('yeah let\'s goooooooooo!');
+}
+
+console.log(sayHelloWorld.name);
+console.log(sayHelloWorld['name']);
+```
+<!-- prettier-ignore-end -->
+
+This may also clarify why we create functions with such long syntax. In the above code, we can now read it as the variable `sayHelloWorld` is assigned to an object. This object is the function `sayHelloWorld`.
+
+### !callout-info
+
+## Foreshadowing: Functions Treated as Objects 😅
+
+JavaScript is not the only language where functions are treated as objects. However, popular and well-known patterns in JavaScript frequently rely on this knowledge! Look forward to seeing functions being treated like objects. 😅
+
+### !end-callout
+
+## Objects Can Hold Functions
+
+This additional topic can illustrate one way that we can treat functions as objects.
+
+When we make objects, we make key-value pairs, where the value can be any object. Values can be numbers, strings, booleans, `null`, arrays, other objects, and _functions_.
+
+Consider [this code](https://replit.com/@adacore/Objects-Demo#index.js):
+
+<!-- prettier-ignore-start -->
+```js
+const shoutMyName = function() {
+  console.log('Kiki!!!');
+}
+
+const employee = {
+  name: 'Kiki',
+  pronouns: 'they/she',
+  greeting: shoutMyName
```

Comment:
Personally, I really wish we would de-emphasize this practice. Yes it works, but if we're making objects with methods, why can't we just use the class keyword like a reasonable person??? When would we actively encourage a student to use this style of creating an object over the much more readable (finally widely supported) class syntax? Even before wide class keyword support, I would have expected prototype classes to be presented over something like this. By this being the first "object-y" thing students see in JS, it predisposes them to think this is how we do classes in JS.